### PR TITLE
Add support for typed arrays

### PR DESCRIPTION
**Pull request by ttencate**

### CODE REVIEW

**Comment by Bromeon** [comment]
bors try

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:206]: This should probably call `check_type()`. 

Even if the layout is the same, the entire invariant of the class is that each element has type `U`.
It isn't unimaginable that future code relies on this, without thinking of this (unchecked yet safe) boundary.

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:487]: Outdated doc? `value` is passed by value.

See main comment regarding design.

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:768]: Maybe add a small doctest example code?

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:316]: The "by reference" is a bit unclear, as the return type is a value. Maybe "referencing the array"?

File: godot-core/src/builtin/meta/class_name.rs
Comment (Bromeon) [inline godot-core/src/builtin/meta/class_name.rs:44]: Maybe an explicit `string_name()` method is clearer? I'm having a bit a hard time deciphering `.into()` conversions all over the code, and this internal API doesn't need to be ergonomic 😉

File: godot-core/src/builtin/variant/impls.rs
Comment (Bromeon) [inline godot-core/src/builtin/variant/impls.rs:221]: Thanks for this clarification! 👍

File: godot/src/lib.rs
Comment (Bromeon) [inline godot/src/lib.rs:39]: Nice! 🙂

**Review by Bromeon — COMMENTED** [review]
Thanks a lot for another great pull request! 😀 

---

**1.** I thought a bit more about the naming, especially about [your comments on Discord](https://discord.com/channels/723850269347283004/1065051136622612590/1068886412088332318):

> I hope that typed arrays will be the majority, so #2 is nicer to work with. It also aligns with gdextension's VariantArray. [...]
> 
> At that boundary, in godot-core/src/gen, I count 87 places where Array is used as argument or return type, and 209 places where TypedArray is used instead. And some of those Array uses might be cases where the Godot devs haven't added extra type annotations yet.

Idiomatic would probably be to make the "right" way (type-safe) easier to use? That is,
```rs
struct Array<T> {...}
type VariantArray = Array<Variant>;
```
I think this might be worth a try. If it doesn't work, we can still do something like `VArray` or even rename everything. 

(Don't bother with all the text re-alignment if you replace identifiers 😬)

--- 

**2.** The current design does mean that in case of type `Variant`, we create extra copies by calling `Variant::to_variant()`. Which I think is fine for now, they're designed to be cheaply copyable, but I thought it would be worth mentioning.


--- 

**3.** The reordering of the methods makes it quite hard to see actual changes, is there any particular reason for that?

**Comment by ttencate** [comment]
> **2.** The current design does mean that in case of type `Variant`, we create extra copies by calling `Variant::to_variant()`. Which I think is fine for now, they're designed to be cheaply copyable, but I thought it would be worth mentioning.

True, I didn't think of that. Agreed that we can live with it for now.

> **3.** The reordering of the methods makes it quite hard to see actual changes, is there any particular reason for that?

There are three impl blocks now: one for `T: VariantMetadata`, one for `T: VariantMetadata + FromVariant` and one for `T: VariantMetadata + ToVariant`. Seemed like the Rusty way to do it. Not sure how useful it is; maybe it would allow e.g. `TypedArray<&str>`, but of course it would be write-only.

If you think it's not worth it, I can easily put it back in a single impl in the original order.

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:653]: It's not too bad to support that actually. Just did it the quick and dirty way for now, but I can DRY it up before we merge this.

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:206]: Note that `assume_typed` is not public. I think we should have a public wrapper like `to_typed()` that _does_ check the type (and returns `Result`).

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:487]: Done.

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:768]: Done.

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:316]: Done.

File: godot-core/src/builtin/meta/class_name.rs
Comment (ttencate) [reply godot-core/src/builtin/meta/class_name.rs:44]: Haha, I did that at first and found it unrusty :) I updated the (only) call site to explicitly call `StringName::from()`.

**Comment by ttencate** [comment]
Looks like `set_typed` was removed from the JSON API, but there's still `array_set_typed` in the header file.

They also [added](https://github.com/godotengine/godot/commit/5909f9f07547895de24fb6965d44c859b69a54a2#diff-29b903ebcf0b26a6b2fa3e617462e06b69b6d5201e1dda76095076949da8a0f8) `array_ref`, which just points the array's data to another array without checking or touching any types. Not sure when you'd want that, but it's good to know it exists.

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [reply godot-core/src/builtin/array.rs:206]: Public/private is generally orthogonal to safe/`unsafe`. This method can be used to subvert the type system in a way that can cause UB. Not in the current implementation (as far as I can see at least), but as soon as any code starts relying on the fact that a `Array<T>` holds indeed variants of type `T`. Which is a generally reasonable assumption*.

Imagine we optimize one of the accesses in the future:
```rs
pub fn get(&self, index: usize) -> T {
    ...
    T::from_variant(variant)
}
```
becomes
```rs
pub fn get(&self, index: usize) -> T {
    ...
    // SAFETY: every Variant in Array<T> has T
    unsafe { T::from_variant_unchecked(variant) }
}
```
Now you can argue that this `unsafe` block introduced UB. But it's not true -- its `// SAFETY` assumption relies on the invariant of the type `Array<T>`. The function that allows this invariant to be violated in the first place is `assume_typed()`, if used wrongly. 

So there are two options:
* `assume_typed()` must be `unsafe`, as its _external_ invariant is that it is always called on arrays with the correct type
* `assume_typed()` verifies that invariant and is thus safe.

---

\* As you brought up, there are edge cases where "`Array<T>` always holds `T` variants" is hard to enforce, but I'd like to try. Even if this turned out to be wrong, I wouldn't see it as a reason to open more doors that can violate invariants.

Comment (ttencate) [reply godot-core/src/builtin/array.rs:206]: Fair point. In an ideal world, I would mark `assume_typed()` as `unsafe` because all current call sites actually do uphold the invariant. However, we don't know what curve balls Godot's typing will throw us in the future, so an extra `check_type()` is the safer route.

Comment (ttencate) [reply godot-core/src/builtin/array.rs:206]: Turns out, we actually need to break the invariant sometimes. `InnerArray::append_array()` for example takes any `Array`, not necessarily a `TypedArray` of the same type as `self`. So I'm going to mark this `unsafe` instead.

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [reply godot-core/src/builtin/array.rs:654]: As you wish. For me, it's also fine to do that later.

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [reply godot-core/src/builtin/array.rs:206]: But why does that make it `unsafe`? The method `assume_typed()` would just panic on `check_type()`, no?

In other words, it's a logic error if we don't allow to append `Array<String>` to `Array<Variant>`, but not a memory safety issue.

File: godot-core/src/builtin/meta/class_name.rs
Comment (Bromeon) [reply godot-core/src/builtin/meta/class_name.rs:44]: Still not a huge fan, but there's more important stuff to argue about 😛

**Comment by Bromeon** [comment]
> There are three impl blocks now: one for `T: VariantMetadata`, one for `T: VariantMetadata + FromVariant` and one for `T: VariantMetadata + ToVariant`. Seemed like the Rusty way to do it. Not sure how useful it is; maybe it would allow e.g. `TypedArray<&str>`, but of course it would be write-only.
> 
> If you think it's not worth it, I can easily put it back in a single impl in the original order.

Ah, good thinking! We may not be able to keep that up in case we try to tighten the invariants, but for this PR it should be fine.

---

> Looks like `set_typed` was removed from the JSON API, but there's still `array_set_typed` in the header file.

Yep, you might have missed [my reply here](https://github.com/godot-rust/gdextension/pull/101#issuecomment-1411790312) 🙂 
This was changed in https://github.com/godotengine/godot/pull/71989, motivation being that it shouldn't be accessible from GDScript.


---

Any further inputs about the name?
From my earlier review:

> **1.** I thought a bit more about the naming, especially about [your comments on Discord](https://discord.com/channels/723850269347283004/1065051136622612590/1068886412088332318):
> 
> > I hope that typed arrays will be the majority, so #2 is nicer to work with. It also aligns with gdextension's VariantArray. [...]
> > At that boundary, in godot-core/src/gen, I count 87 places where Array is used as argument or return type, and 209 places where TypedArray is used instead. And some of those Array uses might be cases where the Godot devs haven't added extra type annotations yet.
> 
> Idiomatic would probably be to make the "right" way (type-safe) easier to use? That is,
> 
> ```rust
> struct Array<T> {...}
> type VariantArray = Array<Variant>;
> ```
> 
> I think this might be worth a try. If it doesn't work, we can still do something like `VArray` or even rename everything.
> 
> (Don't bother with all the text re-alignment if you replace identifiers 😬)
> 
> **2.** The current design does mean that in case of type `Variant`, we create extra copies by calling `Variant::to_variant()`. Which I think is fine for now, they're designed to be cheaply copyable, but I thought it would be worth mentioning.
> 
> **3.** The reordering of the methods makes it quite hard to see actual changes, is there any particular reason for that?

**Comment by ttencate** [comment]
> Any further inputs about the name?

I would like to rename it as you suggested, but would prefer to do that in a separate PR if that's okay.

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:206]: We can't call `check_type()` inside `assume_typed()` because it would panic for a legitimate, but internal, use case: conversion from `TypedArray<T>` to `TypedArray<Variant>` in order to pass it to `InnerArray::append_array()`. You mentioned two options above; we can't use the second, so I went with the first.

**Comment by ttencate** [comment]
Current status: need to add more integration tests; also blocked on #108 which is causing the first few integration tests to fail.

**Comment by Bromeon** [comment]
Sorry for merge conflicts, doctests were broken on `master` and somehow slipped CI. 

> I would like to rename it as you suggested, but would prefer to do that in a separate PR if that's okay.

Fine with me. With your proposal of `UnknownArray`, we might also consider how that would possibly integrate into our naming scheme.

> Current status: need to add more integration tests; also blocked on #108 which is causing the first few integration tests to fail.

I have a branch that fixes the issue, unfortunately it introduces memory leaks elsewhere 🙈 I'll keep debugging.

**Comment by ttencate** [comment]
Alright, this PR has been stewing for long enough, so I'm going to leave generics support in `impl_builtin_traits!` for a future PR.

**Comment by Bromeon** [comment]
Integration tests only have access to a limited set of classes, due to compile-time / dev-cycle reasons.

Would it be possible to use some of the available classes [here](https://github.com/godot-rust/gdextension/blob/master/godot-codegen/src/lib.rs#L258)?

File: godot-codegen/src/class_generator.rs
Comment (Bromeon) [inline godot-codegen/src/class_generator.rs:310]: Why this change? The new method `from_sys_ptr` breaks the link from the lifetime in `&'a outer_class` to `#inner_class<'a>`, and returns [an unbounded lifetime](https://doc.rust-lang.org/nomicon/unbounded-lifetimes.html).

It doesn't seem like you really need it though?

File: godot-codegen/src/util.rs
Comment (Bromeon) [inline godot-codegen/src/util.rs:59]: I'd like to not provide enum defaults yet, especially since `ord: 0` is not necessarily a valid one.

For testing purposes, could you select one of the existing enumerators?

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:206]: Would need a `# Safety` section.

Also, "since it breaks" -> "since it can be used to break".
Hopefully it doesn't _always_ break 😉

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:287]: Why not `to_i64` for `step`?

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:295]: And most importantly, `T` is the same type as `Self` has 🙂 (maybe mention that)

I was first confused, but then realized that the `assume_type()` only does an "upcast" to `TypedArray<Variant>`.

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:303]: Could you replace `try_into().unwrap()` with `as i32`? We're not really benefitting from the check (the difference between `i64` vs. `i32` will never be relevant for our couple of ords), and it's more expressive.

Nice that you use a dedicated `TypeInfo` struct, by the way 🙂

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:321]: I know it's called "set_typed" in Godot, but a setter without a value is very confusing, especially since it's irreversible.

Maybe `become_typed`?

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:328]: The extra parentheses around `interface_fn!(array_set_typed)` should not be necessary (see existing invocations all over the codebase).

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:380]: `T::from_variant`

This occurs quite a few times in this file, maybe search for "try_from_variant".
Now that the `convenience` feature is removed on master, `from_variant` should be accessible everywhere.

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:385]: "If the elements can't be compared" -- when can this happen?

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:798]: Do these pass doctests?
Because the Godot engine needs to be running, which it isn't during `cargo test` 🤔 

If this causes problems, you can mark the sections as `no_run`.

File: godot-core/src/builtin/meta/class_name.rs
Comment (Bromeon) [inline godot-core/src/builtin/meta/class_name.rs:22]: Maybe call it `of`, consistent with [`std::any::TypeId::of`](https://doc.rust-lang.org/std/any/struct.TypeId.html#method.of).

File: itest/rust/src/array_test.rs
Comment (Bromeon) [inline itest/rust/src/array_test.rs:371]: This would theoretically pass if both instance IDs are 0 (`None`).

A stronger test would be to call `instance_id()` directly, as that ensures both sides have a valid instance ID. As a nice bonus, it's also more terse.

**Comment by ttencate** [comment]
Right, I was still using an old command line (not `check.sh itest`) locally. Might be difficult to find a replacement... I'll take a closer look tonight (CEST).

**Review by Bromeon — COMMENTED** [review]
Thanks a lot for the update, and the huge number of tests!

File: godot-codegen/src/class_generator.rs
Comment (ttencate) [reply godot-codegen/src/class_generator.rs:310]: `from_outer` only exists on `Array`, not on `TypedArray<T>`. But I can handle that with another transmute on the caller side.

File: godot-codegen/src/util.rs
Comment (ttencate) [reply godot-codegen/src/util.rs:59]: It's only for bitfields, otherwise there's no sensible way to create an empty bitfield. I'm adding a comment.

Comment (Bromeon) [reply godot-codegen/src/util.rs:59]: Ah, I see!

Wouldn't it be easier to uncomment this? That would be the purpose, I just didn't activate it until we need it:
https://github.com/godot-rust/gdextension/blob/eb5d43b3e021e1eb1b67661d91318660cd54794c/godot-codegen/src/util.rs#L44-L46

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:206]: Added.

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:287]: `to_i64` takes `usize`, not `isize`.

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:295]: I added an explicit `::<Variant>` to make that clearer.

Whether or not `T` is the same type as `Self` does not matter for the safety or unsafety of the upcast.

File: godot-codegen/src/class_generator.rs
Comment (Bromeon) [reply godot-codegen/src/class_generator.rs:310]: Ah, I see. Both unbounded lifetime and transmute is not ideal, but I'd prefer the latter.
Maybe add a `TODO`, then at some point in the future, I can revisit this 😉

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:303]: Done.

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [reply godot-core/src/builtin/array.rs:287]: Ah, again, sorry if I keep asking 🙈

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:321]: `assign_type_to_inner` ok?

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [reply godot-core/src/builtin/array.rs:295]: > I added an explicit `::<Variant>` to make that clearer.

Good idea!

> Whether or not `T` is the same type as `Self` does not matter for the safety or unsafety of the upcast.

True, I was thinking of the `self.as_inner().append_array(other)` rather than the upcast -- but even that is safe, as Godot runtime-checks the types. So you're right, we can leave the description.

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:380]: Done. I didn't mind the extra `unwrap()` since it clearly signals "if this ever triggers, it's a bug".

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [reply godot-core/src/builtin/array.rs:321]: `init_inner_type` sounds great! 👍

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:385]: Copied from upstream. At least ints vs. strings.

File: godot-core/src/builtin/array.rs
Comment (ttencate) [reply godot-core/src/builtin/array.rs:798]: No, and done :)

File: godot-core/src/builtin/meta/class_name.rs
Comment (ttencate) [reply godot-core/src/builtin/meta/class_name.rs:22]: Excellent.

I did this rename because I initially read `ClassName::new::<()>()` as "empty class name", because that's usually what `new` without arguments does. Maybe it was a lack of coffee but this is slightly clearer.

File: itest/rust/src/array_test.rs
Comment (ttencate) [reply itest/rust/src/array_test.rs:371]: Yes, that's better, don't know what I was thinking.

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [reply godot-core/src/builtin/array.rs:380]: > Done. I didn't mind the extra unwrap() since it clearly signals "if this ever triggers, it's a bug".

That's pretty much exactly the rationale behind these panicking methods as well, I just wrote it down today 🙂 

https://github.com/godot-rust/gdextension/blob/eb5d43b3e021e1eb1b67661d91318660cd54794c/godot/src/lib.rs#L72-L75

I agree it may be less obvious to general Rust developers that aren't familiar with the library, but I hope these design documents help explain the trade-offs this library makes 💡

File: godot-core/src/builtin/meta/class_name.rs
Comment (Bromeon) [reply godot-core/src/builtin/meta/class_name.rs:22]: Fully agree, thanks for bringing this up!

File: godot-codegen/src/util.rs
Comment (ttencate) [reply godot-codegen/src/util.rs:59]: Reverted, was only used in an itest that's now gone.

**Review by Bromeon — APPROVED** [review]
Thanks a lot for another great addition!

bors r+

File: godot-core/src/builtin/array.rs
Comment (Bromeon) [inline godot-core/src/builtin/array.rs:317]:
```suggestion
        let variant_type = VariantType::from_sys(self.as_inner().get_typed_builtin() as sys::GDExtensionVariantType);
```

**Review by Bromeon — COMMENTED** [review]
Small suggestion, and hopefully we can additionally integrate https://github.com/godotengine/godot/pull/72926 for type safety. But otherwise we can also make a change that works with current Godot.

**Comment by ttencate** [comment]
Let's deal with that later, once it's merged upstream. Applied your suggestion for now.

**Comment by ttencate** [comment]
Rebased and resolved conflicts.

**Comment by Bromeon** [comment]
Finally 😁

**Pull request merged**
